src/main/java/fr/sncf/osrd/speedcontroller/generators/MarecoAllowanceGenerator.java

```diff
@@ -38,26 +41,83 @@ protected double getTargetTime(double baseTime) {
+     * Compute vf given v1
+     * @param v1 speed limit in MARECO algorithm
+     * @return vf speed when the train starts braking in MARECO algorithm
+     */
+    public double vf(double v1) {
+        var a = schedule.rollingStock.A;
+        var b = schedule.rollingStock.B;
+        var c = schedule.rollingStock.C;
+        double vf = (2 * c * v1 * v1 * v1 + b * v1 * v1) / (3 * c * v1 * v1 + 2 * b * v1 + a);
+        return vf;
+    }
+
+    // we will try to find v1 so that f(v1, vmax) == 0
+    public double newtonsMethod(double v1, double vmax) {
+        return vf(v1) - vmax;
+    }
+
+    /**
+     * First derivative of newtonsMethod
+     * df(v1, vmax)/dv1
+     */
+    public double newtonsMethodDerivative(double v1, double vmax) {
+        var a = schedule.rollingStock.A;
+        var b = schedule.rollingStock.B;
+        var c = schedule.rollingStock.C;
+        var v = (3 * c * v1 * v1 + 2 * b * v1 + a);
+        var dv = (6 * c * v1 + 2 * b);
+        var u = (2 * c * v1 * v1 * v1 + b * v1 * v1) - vmax * v;
+        var du = (6 * c * v1 * v1 + 2 * b * v1) - vmax * dv;
+        return (du * v - dv * u) / (v * v);
+    }
+
     @Override
     protected double getFirstLowEstimate() {
         return 0;
     }
 
+    // get the high boundary for the binary search, corresponding to vf = max
     @Override
     protected double getFirstHighEstimate() {
         double max = 0;
         double position = sectionBegin;
-        while (position < sectionEnd) {
+        double endLocation = sectionEnd;
+        // get max allowed speed
+        while (position < endLocation) {
             double val = SpeedController.getDirective(maxSpeedControllers, position).allowedSpeed;
             if (val > max)
                 max = val;
             position += 1;
         }
-        // TODO find better way to define it
-        return max * 2;
+
+        double tolerance = .000001; // Stop if you're close enough
+        int maxCount = 200; // Maximum number of Newton's method iterations
+        double x = max * 3 / 2; // at high v1 the equation vf = f(v1) tends to be vf = 2*v1/3
+
+        for (int count = 1;
+                //Carry on till we're close, or we've run it 200 times.
+                (Math.abs(newtonsMethod(x, max)) > tolerance) && (count < maxCount);
+                count++)  {
+
+            x = x - newtonsMethod(x, max) / newtonsMethodDerivative(x, max);  //Newtons method.
+        }
+
+        if (Math.abs(newtonsMethod(x, max)) <= tolerance) {
+            return x;
+        } else {
+            return max * 2; // if no value has been found return a high value to have some margin
+        }
+    }
+
+    @Override
+    protected double getFirstGuess() {
+        return this.getFirstHighEstimate() / (1 + value / 100);
     }
 
-    private List<Double> findPositionSameSpeedAsVF(SortedDoubleMap speeds, double vf) {
+    private List<Double> findPositionSameSpeedAsVF(NavigableMap<Double, Double> speeds, double vf) {
```

Comment:
This is backward, we should leave it as a `SortedDoubleMap`. But it doesn't matter much here.